### DISCUSSION TITLE
Porting tinyusb to nrf52840 (USBD_IRQHandler not being called)

### DISCUSSION THREAD

**Discussion by oleh-hordiichuk** [hathach/tinyusb#731] [Q&A] [answered]
Hello, 

I'm porting tinyusb into existing big nrf52840 project (Segger IDE) by replacing nrfx_usbd* implementation with tinyusb. I have disabled all *USBD* modules in sdk_config.h and added tusb_config.h + usb_descriptors.c. Also I implemented the following function in main.c:

`
void USBD_IRQHandler(void)
{
  tud_int_handler(0);
}
`

but it never being called. But as far as I understand tusb_init() has to enable this interrupt by itself, so the function has to be called, e.g. when device is plugged\unplugged from USB. Am I missing something here? Thank you!

P.S.: I understand it's more likely problem in my project setup and the fact that I don't understand exactly how interrupts are working in NRF family devices, but I didn't find anything useful in their documentation. So perhaps someone could point me out where is the problem ;)

**Comment by hathach** [comment, marked as answer]
you need to call the tusb_hal_nrf_power_event() on the power event to trigger enable usb controller. Check out the stock example with nrf board bsp here for hint https://github.com/hathach/tinyusb/blob/master/hw/bsp/nrf/family.c

Note: the power handle is different with and without softdevice enabled, when SD inited/started after usb, you also need to uninit the nrfx_power() and hooked up/ check the power using sd api as well. Here is the code used in adafruit arduino nrf52 https://github.com/adafruit/Adafruit_nRF52_Arduino/blob/master/libraries/Bluefruit52Lib/src/bluefruit.cpp#L67

**Reply by oleh-hordiichuk** [reply]
Thank you so much for the reply!

I checked both links and as we are using SoftDevice, it means we have to call tusb_hal_nrf_power_event in SD_EVT_IRQHandler. However it's already defined in the project and used together with app_scheduler module:

```c
void nrf_sdh_evts_poll(void)
{
    nrf_section_iter_t iter;

    // Notify observers about pending SoftDevice event.
    for (nrf_section_iter_init(&iter, &sdh_stack_observers);
         nrf_section_iter_get(&iter) != NULL;
         nrf_section_iter_next(&iter))
    {
        nrf_sdh_stack_observer_t    * p_observer;
        nrf_sdh_stack_evt_handler_t   handler;

        p_observer = (nrf_sdh_stack_observer_t *) nrf_section_iter_get(&iter);
        handler    = p_observer->handler;

        handler(p_observer->p_context);
    }
}

static void appsh_events_poll(void * p_event_data, uint16_t event_size)
{
    UNUSED_PARAMETER(p_event_data);
    UNUSED_PARAMETER(event_size);

    nrf_sdh_evts_poll();
}

void SD_EVT_IRQHandler(void)
{
    ret_code_t ret_code = app_sched_event_put(NULL, 0, appsh_events_poll);
    APP_ERROR_CHECK(ret_code);
}
```

So all SoftDevice events are being executed in main(). Therefore:

1) Will it be correct if I simply subscribe to SoftDevice as nrf_sdh_soc_evt_observer_t and then call tusb_hal_nrf_power_event from there? The nrf_sdh_soc_evt_observer_t handler contains evt_id parameter obtained from sd_evt_get();
2) Is it ok if tusb_hal_nrf_power_event will be called from main, not interrupt?

Thank you in advance!

**Reply by hathach** [reply]
1. I don't use nrf_sdh() so I don't know the answer, but what you looked for it already in Adafruit_nRF52_Arduino link above just without sdh()
2. tusb_hal_nrf_power_event() is supposed to be called in main/thread mode.

**Reply by oleh-hordiichuk** [reply]
Thank you for help! I've implemented it with nrf_sdh and works as it should!

**Comment by chenyingchun0312** [comment]
Hi, @oleh-hordiichuk :
     You said "I'm porting tinyusb into existing big nrf52840 project (Segger IDE) by replacing nrfx_usbd* implementation with tinyusb.", Does this mean， you want to use nrfx_usbd* apis replace with  `https://github.com/hathach/tinyusb/blob/master/src/portable/nordic/nrf5x/dcd_nrf5x.c` register implemation? 

    As currenttly `https://github.com/hathach/tinyusb/blob/master/src/portable/nordic/nrf5x/dcd_nrf5x.c` implemation, which do not use nrfx_usbd* api, the code is difficult to read.

    Waitting for your reply，thank you !
